internal/cmd/versions.go

```diff
@@ -15,24 +15,104 @@
 package cmd
 
 import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
 	"github.com/urfave/cli/v2"
 
 	"github.com/tetratelabs/getenvoy/internal/envoy"
 	"github.com/tetratelabs/getenvoy/internal/globals"
+	"github.com/tetratelabs/getenvoy/internal/version"
 )
 
 // NewVersionsCmd returns command that lists available Envoy versions for the current platform.
 func NewVersionsCmd(o *globals.GlobalOpts) *cli.Command {
 	return &cli.Command{
 		Name:  "versions",
-		Usage: "List available Envoy versions",
+		Usage: "List Envoy versions",
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:    "all",
+				Aliases: []string{"a"},
+				Usage:   "Show all versions including ones not yet installed",
+			}},
 		Action: func(c *cli.Context) error {
-			m, err := envoy.GetEnvoyVersions(o.EnvoyVersionsURL, o.UserAgent)
+			rows, err := getInstalledVersions(o.HomeDir)
 			if err != nil {
 				return err
 			}
-			envoy.PrintVersions(m, globals.CurrentPlatform, c.App.Writer)
+			if c.Bool("all") {
+				if ev, err := envoy.GetEnvoyVersions(o.EnvoyVersionsURL, o.UserAgent); err != nil {
+					return err
+				} else if err := addAvailableVersions(&rows, ev.Versions, globals.CurrentPlatform); err != nil {
+					return err
+				}
+			}
+
+			if len(rows) == 0 {
+				fmt.Fprintln(c.App.Writer, "No Envoy versions, yet") //nolint
+				return nil
+			}
+
+			// Sort so that new release dates appear first and on conflict choosing the higher version
+			sort.Slice(rows, func(i, j int) bool {
+				if rows[i].releaseDate == rows[j].releaseDate {
+					return rows[i].version > rows[j].version
+				}
+				return rows[i].releaseDate > rows[j].releaseDate
+			})
+
+			// This doesn't use tabwriter because the columns are likely to remain the same width.
+			fmt.Fprintln(c.App.Writer, "VERSION\tRELEASE_DATE") //nolint
+			for _, vr := range rows {                           //nolint:gocritic
+				fmt.Fprintf(c.App.Writer, "%s\t%s\n", vr.version, vr.releaseDate) //nolint
+			}
 			return nil
 		},
 	}
 }
+
+type versionReleaseDate struct {
+	// version ex "1.15.5"
+	version string
+	// releaseDate ex "2021-05-11"
+	releaseDate string
+}
+
+func getInstalledVersions(homeDir string) ([]versionReleaseDate, error) {
+	var rows []versionReleaseDate
+	files, err := os.ReadDir(filepath.Join(homeDir, "versions"))
+	if os.IsNotExist(err) {
+		return rows, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if i, err := f.Info(); f.IsDir() && err == nil {
+			rows = append(rows, versionReleaseDate{f.Name(), i.ModTime().Format("2006-01-02")})
+		}
+	}
+	return rows, nil
+}
+
+// addAvailableVersions adds remote Envoy versions valid for this platform to "vr", if they don't already exist
+func addAvailableVersions(rows *[]versionReleaseDate, remote map[string]version.EnvoyVersion, p string) error {
+	existingVersions := make(map[string]bool)
```

Comment:
formerly I made the input a map, but it is cleaner to declare a temporary map as we get the field names everywhere else now.

Reply:
we in general use `map[string]struct{}` for this purpose (a bit (literally) less than map[string]bool)

Reply:
I didn't know that! As far as I can research, this practice is to save memory from allocating a boolean.

I might push back on this a little as memory isn't an issue for the versions list and the call site becomes harder to use.

Ex. `!existingVersions[k]` is easier to use than `_, ok := existingVersions[k]; ok` especially in the condition below which is already doing a boolean chain once..

ok to leave it?

Reply:
sg!

Reply:
Even if the bit to be safe isn't a big deal, I would say it is a common practice and easier to read in that way because you know what to expect.